jiant/utils/utils.py

```diff
@@ -361,15 +401,38 @@ def get_elmo_mixing_weights(text_field_embedder, task=None):
     return params
 
 
-def get_batch_size(batch, keyword="input"):
+def format_output(obj, cuda_devices):
+    """
+    Format output based on whether model is using DataParallel or not. 
+    DataParallel necessitates objects to be gathered into GPU:0 to have 
+    dimension 0. 
+    This function will be used for scalar outputs of model forwards 
+    such as loss and n_exs.
+    """
+    if isinstance(cuda_devices, list):
+        if not isinstance(obj, torch.Tensor):
+            obj = torch.tensor(obj).cuda()
+        return obj.unsqueeze(0)
+    else:
+        return obj
+
+
+def uses_cuda(cuda_devices):
+    use_cuda = 1
+    if isinstance(cuda_devices, list) or isinstance(cuda_devices, int) and cuda_devices >= 0:
+        return use_cuda
+    return 0
+
+
+def get_batch_size(batch, cuda_devices, keyword="input"):
```

Comment:
if we're doing distributed training, does it but the same batch size on all devices (so that we'd need to multiply by number of devices here), or is it splitting the batch across devices ?

Reply:
It splits up the batch into equal parts to all devices. However, we don't have to multiply because the get_output_attribute sums the n_exes together from all batches.